src/mon/OSDMonitor.cc

```diff
@@ -1438,8 +1449,353 @@ void OSDMonitor::encode_trim_extra(MonitorDBStore::TransactionRef tx,
   bufferlist bl;
   get_version_full(first, bl);
   put_version_full(tx, first, bl);
+
+  if (has_osdmap_manifest &&
+      first > osdmap_manifest.get_first_pinned()) {
+    _prune_update_trimmed(tx, first);
+  }
+}
+
+
+/* full osdmap prune
+ *
+ * for more information, please refer to doc/dev/mon-osdmap-prune.rst
+ */
+
+void OSDMonitor::update_osdmap_manifest()
+{
+  bool store_has_manifest =
+    mon->store->exists(get_service_name(), "full_manifest");
+
+  if (!store_has_manifest) {
+    if (!has_osdmap_manifest) {
+      return;
+    }
+
+    dout(20) << __func__
+             << " dropping osdmap manifest from memory." << dendl;
+
+    osdmap_manifest = osdmap_manifest_t();
+    has_osdmap_manifest = false;
+    return;
+  }
+
+  dout(20) << __func__
+           << " osdmap manifest detected in store; reload." << dendl;
+
+  bufferlist manifest_bl;
+  int r = get_value("full_manifest", manifest_bl);
+  if (r < 0) {
+    derr << __func__ << " unable to read osdmap version manifest" << dendl;
+    ceph_assert(0 == "error reading manifest");
+  }
+  osdmap_manifest.decode(manifest_bl);
+  has_osdmap_manifest = true;
+
+  dout(10) << __func__ << " store osdmap manifest last pinned "
+           << osdmap_manifest.get_last_pinned()
+           << " last pruned " << osdmap_manifest.get_last_pruned()
+           << dendl;
+}
+
+bool OSDMonitor::is_prune_ongoing() const
+{
+  if (!has_osdmap_manifest)
+    return false;
+
+  version_t next = osdmap_manifest.get_next_prunable();
+
+  if (g_conf->get_val<bool>("mon_debug_extra_checks") && next > 0) {
+
+    version_t upper = osdmap_manifest.get_upper_closest_pinned(next);
+    version_t lower = osdmap_manifest.get_lower_closest_pinned(next);
+
+    ceph_assert(upper > 0);
+    ceph_assert(lower > 0);
+    ceph_assert(upper <= osdmap_manifest.get_last_pinned());
+    ceph_assert(lower >= osdmap_manifest.get_first_pinned());
+    ceph_assert(lower < upper);
+  }
+
+  return (next > 0);
+}
+
+bool OSDMonitor::should_prune() const
+{
+  version_t first = get_first_committed();
```

Comment:
i'd suggest move the definition of `first`, `last`, `min_osdmap_epochs`, `prune_to` and `last_pruned` after the following `if` block.

Reply:
yeah

Reply:
@liewegas @tchaikov sorry, where to? Somehow, it looks like these comments have been kept, even though the code itself has changed (heck, github tells me the commit this review is on has changed). Not only there's no `following if block` atm (it's a bunch of if-elseif-else that actually depend on these variables to be declared and initialized), but everything pertaining to `last_pruned` has been dropped from the current algorithm.